lib/gems/pending/util/extensions/miq-module.rb

```diff
@@ -26,15 +26,17 @@ def cache_with_timeout(method, timeout = nil, &block)
         cache = $miq_cache_with_timeout[key]
 
         old_timeout = cache[:timeout]
-        cache.clear if force_reload || (old_timeout && Time.now.utc > old_timeout)
-        break cache[:value] unless cache.empty?
+        timeout_expired = !!old_timeout && (Time.now.utc > old_timeout)
+        break cache[:value] if (!force_reload && !timeout_expired && cache.key?(:value))
```

Comment:
Yeah, I think this code is doing exactly what you describe...

Basically, we only return from the cache:
If it's not force reload, and, the timeout hasn't expired, and the cache has the value key.